### ISSUE TITLE
Cannot verify PredicateType `cyclonedx` generated with previous versions of cosign, or vice versa.

### ISSUE THREAD

**Issue by vaikas** [sigstore/cosign#2596]
**Description**

The predicate URL for cyclonedx was changed here:
https://github.com/in-toto/in-toto-golang/commit/86e515c6e15f0d9bc2cc9cb8253165f8176f6ce6

And the CLI options allow for convenience fields (in this case `cyclonedx`) which are keyed off of here:
https://github.com/sigstore/cosign/blob/main/cmd/cosign/cli/options/predicate.go#L40

So, depending on when your cosign was built, you're going to get `cyclonedx` URI mapping to either:
`https://cyclonedx.org/schema` or `https://cyclonedx.org/schema`

And then when handling attestation here, the predicate URI has now changed from `schema` to `bom`, it will not be considered valid:
https://github.com/sigstore/cosign/blob/main/pkg/policy/attestation.go#L75

So, if you attest with say cosign 1.13.1 using `--type=cyclonedx` but use the head (soon to be v2) with the `--type=cyclonedx` it will fail, and vice versa.

You can of course use the full URL, but one could argue that it dilutes the value of the convenience types.

**Version**

**Comment by hectorj2f** [comment]
Yes, I agree :/ . We should handle both URIs when type cyclonedx is specified.

**Comment by znewman01** [comment]
See also https://github.com/sigstore/cosign/issues/2612

**Comment by mattmoor** [comment]
@znewman01 @priyawadhwa how contentious would it be to eliminate these shortnames for cosign v2?

They sure cause me endless grief :)

**Comment by AAH20** [comment]
I can contribute a focused fix for this against the current development branch.

The compatibility rule I propose is intentionally narrow:

- generation with `--type cyclonedx` continues to emit the canonical `https://cyclonedx.org/bom` predicate type;
- verification/filtering requested through the `cyclonedx` short name accepts exactly:
  - `https://cyclonedx.org/bom`
  - `https://cyclonedx.org/schema`
- a caller supplying either URI explicitly gets exact matching, not alias expansion;
- no prefix, suffix, host-only, redirect, or case-insensitive URI matching;
- policy consumers continue to receive the predicate URI that was actually present in the verified statement.

This preserves strict behavior for custom predicate URIs while allowing existing CycloneDX evidence to survive a Cosign upgrade without being regenerated and resigned.

## Proposed implementation and tests

- keep `PredicateTypeMap[cyclonedx]` canonical for statement generation;
- add one small predicate-matching helper for verification/policy filtering rather than changing the global map into an ambiguous one-to-many signing map;
- exercise the helper through `AttestationToPayloadJSON` and the relevant CLI verification path;
- verify canonical and legacy CycloneDX predicates in both supported bundle formats;
- verify that explicit canonical and legacy URI requests remain exact;
- reject lookalike domains, URI suffixes, query additions, fragments, and malformed values;
- retain the observed predicate type in mismatch and success results;
- add a CycloneDX 1.7 CBOM fixture containing classical and ML-KEM/ML-DSA cryptographic assets, relationships, and occurrences to prove that current/future predicate fields survive attestation generation and verification;
- retain a subject-digest mismatch regression test so accepting the legacy predicate alias cannot weaken subject binding.

I would keep schema validation out of scope: Cosign should preserve and authenticate the CycloneDX predicate, while CycloneDX tooling validates the BOM itself.

Two questions before implementation:

1. Should successful verification of the legacy `/schema` URI emit a warning, or remain silent for compatibility?
2. Would you prefer the CycloneDX 1.7 CBOM fixture to replace the existing 1.4 fixture, or be added as a separate forward-compatibility case?

Once those choices are clear, I can post the file-level change list and start with the failing compatibility tests.

Disclosure: I maintain related open-source PQC/CBOM work and work through [A2Z SOC](https://a2zsoc.com/), which provides scoped cryptographic-inventory and signed-evidence readiness assessments. The Cosign implementation, tests, fixtures, and documentation would remain vendor-neutral and contain no service promotion.
